jvm-packages/xgboost4j-spark-gpu/src/main/scala/ml/dmlc/xgboost4j/scala/rapids/spark/GpuPreXGBoost.scala

```diff
@@ -0,0 +1,572 @@
+  }
+
+  /**
+   * Transform schema
+   *
+   * @param est supporting XGBoostClassifier/XGBoostClassificationModel and
+   *                 XGBoostRegressor/XGBoostRegressionModel
+   * @param schema   the input schema
+   * @return the transformed schema
+   */
+  override def transformSchema(
+      est: XGBoostEstimatorCommon,
+      schema: StructType): StructType = {
+
+    val fit = est match {
+      case _: XGBoostClassifier | _: XGBoostRegressor => true
+      case _ => false
+    }
+
+    val Seq(label, weight, margin) = GpuUtils.getColumnNames(est)(est.labelCol, est.weightCol,
+      est.baseMarginCol)
+
+    GpuUtils.validateSchema(schema, est.getFeaturesCols, label, weight, margin, fit)
+  }
+
+  /**
+   * Repartition all the Columnar Dataset (training and evaluation) to nWorkers,
+   * and assemble them into a map
+   */
+  private def prepareInputData(
+      trainingData: ColumnDataBatch,
+      evalSetsMap: Map[String, ColumnDataBatch],
+      nWorkers: Int,
+      isCacheData: Boolean): Map[String, ColumnDataBatch] = {
+    // Cache is not supported
+    if (isCacheData) {
+      logger.warn("Dataset cache is not support for GPU pipeline!")
+    }
+
+    (Map(TRAIN_NAME -> trainingData) ++ evalSetsMap).map {
+      case (name, colData) =>
+        // No light cost way to get number of partitions from DataFrame, so always repartition
+        val newDF = colData.groupColName
+          .map(gn => repartitionForGroup(gn, colData.rawDF, nWorkers))
+          .getOrElse(colData.rawDF.repartition(nWorkers))
+        name -> ColumnDataBatch(newDF, colData.colIndices, colData.groupColName)
+    }
+  }
+
+  private def repartitionForGroup(
+      groupName: String,
+      dataFrame: DataFrame,
+      nWorkers: Int): DataFrame = {
+    // Group the data first
+    logger.info("Start groupBy for LTR")
+    val schema = dataFrame.schema
+    val groupedDF = dataFrame
+      .groupBy(groupName)
+      .agg(collect_list(struct(schema.fieldNames.map(col): _*)) as "list")
+
+    implicit val encoder = RowEncoder(schema)
+    // Expand the grouped rows after repartition
+    groupedDF.repartition(nWorkers).mapPartitions(iter => {
+      new Iterator[Row] {
+        var iterInRow: Iterator[Any] = Iterator.empty
+
+        override def hasNext: Boolean = {
+          if (iter.hasNext && !iterInRow.hasNext) {
+            // the first is groupId, second is list
+            iterInRow = iter.next.getSeq(1).iterator
+          }
+          iterInRow.hasNext
+        }
+
+        override def next(): Row = {
+          iterInRow.next.asInstanceOf[Row]
+        }
+      }
+    })
+  }
+
+  private def buildRDDWatches(
+      dataMap: Map[String, ColumnDataBatch],
+      xgbExeParams: XGBoostExecutionParams,
+      noEvalSet: Boolean): RDD[Watches] = {
+
+    val sc = dataMap(TRAIN_NAME).rawDF.sparkSession.sparkContext
+    val maxBin = xgbExeParams.toMap.getOrElse("max_bin", 256).asInstanceOf[Int]
```

Comment:
Can we lower this configuration into C++ or set a default parameter on the API that's more visible?

Reply:
Actually, XGBoost java has set the default value to 256. Here, it's just a safer way to get max_bin.

Reply:
If the parameter is not optional then I think this safety is not necessary.  We should aim for removing any duplicated logic.

Reply:
I don't think this is needed to be removed. but if you insist, I can get rid of it in the following PR